hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+        uint16_t directed      : 1;     /**< Directed advertising event type. */
+        uint16_t extended_pdu  : 1;     /**< Received an extended advertising set. */
+    } type;
+    uint16_t adv_data_len;
+    uint8_t* adv_data;
+    uint16_t sr_data_len;
+    uint8_t* sr_data;
+    hal_ble_addr_t peer_addr;
+} hal_ble_gap_on_scan_result_evt_t;
+
+typedef struct {
+    uint8_t version;
+    uint8_t role;
+    uint16_t conn_handle;
+    uint16_t conn_interval;             /**< Connection Interval in 1.25 ms units.*/
+    uint16_t slave_latency;             /**< Slave Latency in number of connection events.*/
+    uint16_t conn_sup_timeout;          /**< Connection Supervision Timeout in 10 ms units.*/
+    hal_ble_addr_t peer_addr;
+} hal_ble_gap_on_connected_evt_t;
+
+typedef struct {
+    uint8_t version;
+    uint8_t reason;
+    uint16_t conn_handle;
+} hal_ble_gap_on_disconnected_evt_t;
 
-// API version
-#define BLE_API_VERSION 1
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint16_t conn_interval;             /**< Connection Interval in 1.25 ms units.*/
+    uint16_t slave_latency;             /**< Slave Latency in number of connection events.*/
+    uint16_t conn_sup_timeout;          /**< Connection Supervision Timeout in 10 ms units.*/
+} hal_ble_gap_on_conn_params_evt_t;
 
-// Default maximum size of an ATT packet in bytes (ATT_MTU)
-#define BLE_MIN_ATT_MTU_SIZE 23
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint16_t att_mtu_size;
+} hal_ble_gatt_on_params_updated_evt_t;
 
-// Size of the ATT opcode field in bytes
-#define BLE_ATT_OPCODE_SIZE 1
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint8_t  count;
+    hal_ble_svc_t* services;
+} hal_ble_gattc_on_svc_disc_evt_t;
 
-// Size of the ATT handle field in bytes
-#define BLE_ATT_HANDLE_SIZE 2
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint8_t count;
+    hal_ble_char_t* characteristics;
+} hal_ble_gattc_on_char_disc_evt_t;
 
-// Minimum and maximum number of bytes that can be sent in a single write command, read response,
-// notification or indication packet
-#define BLE_MIN_ATTR_VALUE_PACKET_SIZE (BLE_MIN_ATT_MTU_SIZE - BLE_ATT_OPCODE_SIZE - BLE_ATT_HANDLE_SIZE)
-#define BLE_MAX_ATTR_VALUE_PACKET_SIZE (BLE_MAX_ATT_MTU_SIZE - BLE_ATT_OPCODE_SIZE - BLE_ATT_HANDLE_SIZE)
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint16_t attr_handle;
+    uint16_t offset;
+    uint16_t data_len;
+    uint8_t* data;
+} hal_ble_gatt_on_data_evt_t;
+
+typedef enum {
+    BLE_EVT_UNKNOWN,
+    BLE_EVT_ADV_STOPPED,
+    BLE_EVT_SCAN_RESULT,
+    BLE_EVT_CONNECTED,
+    BLE_EVT_DISCONNECTED,
+    BLE_EVT_CONN_PARAMS_UPDATED,
+    BLE_EVT_GATT_PARAMS_UPDATED,
+    BLE_EVT_SVC_DISCOVERED,
+    BLE_EVT_CHAR_DISCOVERED,
+    BLE_EVT_DATA_WRITTEN,
+    BLE_EVT_DATA_NOTIFIED
+} hal_ble_evts_type_t;
+
+typedef struct {
+    hal_ble_evts_type_t type;
```

Comment:
It's better to use integer types instead of enums in dynamically linked code, as the underlying type of an enum may be inadvertently changed

Reply:
See https://github.com/particle-iot/device-os/pull/1740#discussion_r279363737